[FLINK-27571][benchmark] Recognize 'less is better' benchmarks in regression detection script

fredia opened the pull request:
1. Add the script that updates fields for codespeed_benchmark table. After updating, the Y axis  would become “ms/op (less is better)”, e.g. [rescaleHeap](http://codespeed.dak8s.net:8000/timeline/#/?ben=rescaleHeap.RESCALE_IN&extr=on&quarts=on&equid=off&env=2&revs=200&exe=1,3,5,6,8,9)
2. Update save_jmh_results.py , If a new benchmark is introduced, the `lessisbetter` and `units` fields of results would be determined by a new benchmark instead of hard-code.
3. Update regression_report.py to detect the "less is better" regression.

fredia commented:
Hi @pnowojski, could you please help take a look?

rkhachatryan commented on regression_report.py:
```diff
@@ -131,8 +133,12 @@ def checkBenchmark(args, exe, benchmark):
-    if median_trend < args.medianTrendThreshold:
-        # regression detected, print warning
+
+    # regression detected, print warning
+    if median_trend < args.medianTrendThreshold and lessIsbBetter is False:
+        print "<%s|%s> median=%s recent_median=%s" % (urlToBenchmark, benchmark, median, recent_median)
+
+    if median_trend > (-1 * args.medianTrendThreshold) and lessIsbBetter:
```
These two lines don't seem very readable to me.
How about extracting it into a function e.g. `isThresholdReached` returning a boolean?
That would also eliminate code duplication.

fredia replied:
👍Good point, extracted it into `isThresholdReached`.

rkhachatryan commented on update_benchmark_fileds.py:
```diff
@@ -0,0 +1,56 @@
+    conn = sqlite3.connect('data.db')
+    c = conn.cursor()
+    for name in lessisbetter_name:
+        sql = "update codespeed_benchmark set lessisbetter=?,units=? where name=?;"
+        cursor = c.execute(sql, (1, 'ms/op', name))
+        conn.commit()
```
Not sure if this script (`update_benchmark_fileds.py`) should be added to git because it was a one-off task IIUC.
But if it should, then its purpose and some instructions on how to use it would be great.

fredia replied:
👍Agreed, when `save_jmh_result.py` changed, this script is no longer needed, deleted it.

rkhachatryan reviewed (commented):
Thanks for the PR @fredia !
In general, LGTM.
I've left some comments, PTAL.

XComp commented:
@fredia what's the state on this PR? Looks like @rkhachatryan comments still wait for some reply?

fredia commented:
Thanks for the reviewing and advice, and sorry for the late response, the update of codespeed_benchmark table at Jenkins master was done,  and the update of `save_jmh_result.py` and `regression_report.py` need to be reviewed.

pnowojski commented:
Have you noted that since a couple of weeks, codespeed is reporting regression/improvement colors incorrectly? http://codespeed.dak8s.net:8000/changes/ today for example 18% improvement in the sortedOneInput is reported as a regression. Does it have something to do with the lessisbetter changes? Do you know if something has changed in this area? I'm not sure, but I vaguely remember that there might have been a bug in the codespeed, that if even one benchmark is marked as "lessisbetter", all benchmarks will be treated this way.

fredia commented:
@pnowojski  Thanks for your information, I overlooked this when updating `codespeed_benchmark` table. The colors are controlled by the[ group](https://github.com/tobami/codespeed/blob/263860bc298fd970c8466b3161de386582e4f801/codespeed/models.py#L444) of `units_title`, the code assumes that the same `units_title` has the same `lessIsBetter` flags, then return the table for distinct `units_title`.

All the `units_title`'s were the same before, after I updated the `units_title`'s of lessIsBetter benchmark, the color became normal:
![image](https://user-images.githubusercontent.com/18653940/216885235-ee94b22f-0fb9-46b3-9061-fbf486766aa3.png)

But for the default trend, the colors are still wrong due to the existence of the [cache](https://github.com/tobami/codespeed/blob/263860bc298fd970c8466b3161de386582e4f801/codespeed/models.py#L415)
![image](https://user-images.githubusercontent.com/18653940/216885914-56190f80-06fb-41f6-9f02-bb69ee935693.png).

pnowojski commented:
Nice find.

What's the issue with the cache? I've checked the report visible on your second screen and it seems to be working properly.  

Anyway, I presume that if there is an issue with cache, it will fix itself soon.

fredia commented:
Thanks for your feedback and advice, I have updated the `save_jmh_result.py`.
The issue with the cache of trend=10(default trend) still exists, e.g.
1. There is only one table in the response of trend=10: http://codespeed.dak8s.net:8000/changes/table/?tre=10&rev=7326addbb3f-9fa36c4-1675623864&exe=1&env=2 
2. There are **two tables** in the response of trend=20: http://codespeed.dak8s.net:8000/changes/table/?tre=20&rev=7326addbb3f-9fa36c4-1675623864&exe=1&env=2

Because the "cache" for a specific `(revision, executable, environment)` is [stored in the database](https://github.com/tobami/codespeed/blob/263860bc298fd970c8466b3161de386582e4f801/codespeed/models.py#L261), it will not change automatically unless reuploading or updating the database manually, I'd like to manually update the database if necessary.

pnowojski commented:
Ahhh, I see it now. By correcting the `units_title`, you have fixed the colouring of the `more is better` benchmarks, but all of the reports that have been generated before that change, are still treating `less is better` as if had been `more is better`. Right?

1. it seems like all the reports generated after your `units_title` fix will be already fixed and working properly
2. if need, we can easily regenerate past reports. In `http://codespeed.dak8s.net:8000/admin/codespeed/report/` we can click to recalculate the reports. I have done that for the last 100 reports. I don't think we need to fix older reports.

fredia commented:
Right, the reports that have been generated before correcting the `units_title` are still treated as `less is better`.
Thanks for your information,  regenerating past reports by codespeed admin is a more convenient method.👍
I changed the `regression_report.py` and `save_jmh_result.py` finally, could you please take a look in your free time?

pnowojski reviewed (approved):
LGTM % nit

fredia commented:
Thanks for your review, I dropped the comment and squashed the commits.

The current notifications are as follows:
> [mapRebalanceMapSink.F27_UNBOUNDED](http://codespeed.dak8s.net:8000/timeline/#/?exe=1&ben=mapRebalanceMapSink.F27_UNBOUNDED&extr=on&quarts=on&equid=off&env=2&revs=200) median=17231.7398765 recent_median=16483.040695
[~~calculateRegionToRestart.BATCH~~](http://codespeed.dak8s.net:8000/timeline/#/?exe=8&ben=calculateRegionToRestart.BATCH&extr=on&quarts=on&equid=off&env=2&revs=200) median=12.827356 recent_median=11.9279615
[~~partitionRelease.BATCH~~](http://codespeed.dak8s.net:8000/timeline/#/?exe=8&ben=partitionRelease.BATCH&extr=on&quarts=on&equid=off&env=2&revs=200) median=23.112702 recent_median=21.378349
[checkpointMultiInput](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=checkpointMultiInput&extr=on&quarts=on&equid=off&env=2&revs=200) median=2.5996435 recent_median=2.490419
[checkpointSingleInput.UNALIGNED](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=checkpointSingleInput.UNALIGNED&extr=on&quarts=on&equid=off&env=2&revs=200) median=338.060806 recent_median=308.487419
[checkpointSingleInput.UNALIGNED_1](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=checkpointSingleInput.UNALIGNED_1&extr=on&quarts=on&equid=off&env=2&revs=200) median=215.0116175 recent_median=192.939931
[fireProcessingTimers](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=fireProcessingTimers&extr=on&quarts=on&equid=off&env=2&revs=200) median=50.6221895 recent_median=47.733433
[globalWindow](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=globalWindow&extr=on&quarts=on&equid=off&env=2&revs=200) median=5459.689767 recent_median=4600.3655335
[stateBackends.FS](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=stateBackends.FS&extr=on&quarts=on&equid=off&env=2&revs=200) median=3269.6932005 recent_median=3084.736179
[stateBackends.FS_ASYNC](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=stateBackends.FS_ASYNC&extr=on&quarts=on&equid=off&env=2&revs=200) median=3261.0470545 recent_median=2894.1319305
[stateBackends.MEMORY](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=stateBackends.MEMORY&extr=on&quarts=on&equid=off&env=2&revs=200) median=3220.0443795 recent_median=2836.4578395
[tupleKeyBy](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=tupleKeyBy&extr=on&quarts=on&equid=off&env=2&revs=200) median=4173.447853 recent_median=3483.1905105

After this PR, `calculateRegionToRestart.BATCH` and `partitionRelease.BATCH` will disappear,  `buildTopology.STREAMING`, `createScheduler.STREAMING` and `deployDownstreamTasks.BATCH` will be added:

> [mapRebalanceMapSink.F27_UNBOUNDED](http://codespeed.dak8s.net:8000/timeline/#/?exe=1&ben=mapRebalanceMapSink.F27_UNBOUNDED&extr=on&quarts=on&equid=off&env=2&revs=200) median=17146.759315 recent_median=16434.359172
[checkpointSingleInput.UNALIGNED](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=checkpointSingleInput.UNALIGNED&extr=on&quarts=on&equid=off&env=2&revs=200) median=337.6920935 recent_median=308.487419
[checkpointSingleInput.UNALIGNED_1](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=checkpointSingleInput.UNALIGNED_1&extr=on&quarts=on&equid=off&env=2&revs=200) median=214.956786 recent_median=193.5968605
[fireProcessingTimers](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=fireProcessingTimers&extr=on&quarts=on&equid=off&env=2&revs=200) median=50.5016795 recent_median=48.207456
[globalWindow](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=globalWindow&extr=on&quarts=on&equid=off&env=2&revs=200) median=5470.4972005 recent_median=4600.3655335
[stateBackends.FS](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=stateBackends.FS&extr=on&quarts=on&equid=off&env=2&revs=200) median=3276.4542755 recent_median=2886.6052725
[stateBackends.FS_ASYNC](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=stateBackends.FS_ASYNC&extr=on&quarts=on&equid=off&env=2&revs=200) median=3261.0470545 recent_median=2906.11805
[stateBackends.MEMORY](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=stateBackends.MEMORY&extr=on&quarts=on&equid=off&env=2&revs=200) median=3224.299765 recent_median=2836.4578395
[tupleKeyBy](http://codespeed.dak8s.net:8000/timeline/#/?exe=6&ben=tupleKeyBy&extr=on&quarts=on&equid=off&env=2&revs=200) median=4173.447853 recent_median=3483.1905105
[**buildTopology.STREAMING**](http://codespeed.dak8s.net:8000/timeline/#/?exe=8&ben=buildTopology.STREAMING&extr=on&quarts=on&equid=off&env=2&revs=200) **median=116.1662695 recent_median=121.501681**
[**createScheduler.STREAMING**](http://codespeed.dak8s.net:8000/timeline/#/?exe=8&ben=createScheduler.STREAMING&extr=on&quarts=on&equid=off&env=2&revs=200) **median=283.1718575 recent_median=295.282613**
[**deployDownstreamTasks.BATCH**](http://codespeed.dak8s.net:8000/timeline/#/?exe=8&ben=deployDownstreamTasks.BATCH&extr=on&quarts=on&equid=off&env=2&revs=200) **median=72.3862015 recent_median=90.633917**
